core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableProject.java

```diff
@@ -86,4 +100,52 @@ public Result implement(EnumerableRelImplementor implementor, Prefer pref) {
     // EnumerableCalcRel is always better
     throw new UnsupportedOperationException();
   }
+
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> passThroughTraits(
+      RelTraitSet required) {
+    RelCollation collation = required.getCollation();
+    if (collation == null || collation.getFieldCollations().size() == 0) {
+      return null;
+    }
+    final Mappings.TargetMapping map =
+        RelOptUtil.permutationIgnoreCast(
+            getProjects(), getInput().getRowType());
+
+    // Determine mapping between project input and output fields. If sort
+    // relies on non-trivial expressions, we can't push.
+    for (RelFieldCollation fc : collation.getFieldCollations()) {
+      if (map.getTargetOpt(fc.getFieldIndex()) < 0) {
+        return null;
+      }
+      final RexNode node = getProjects().get(fc.getFieldIndex());
+      if (node.isA(SqlKind.CAST)) {
+        // Check whether it is a monotonic preserving cast, otherwise we cannot push
+        final RexCall cast = (RexCall) node;
+        RelFieldCollation newFc = Objects.requireNonNull(RexUtil.apply(map, fc));
+        final RexCallBinding binding =
+            RexCallBinding.create(getCluster().getTypeFactory(), cast,
+                ImmutableList.of(RelCollations.of(newFc)));
+        if (cast.getOperator().getMonotonicity(binding) == SqlMonotonicity.NOT_MONOTONIC) {
+          return null;
+        }
+      }
+    }
+
+    final RelCollation newCollation = collation.apply(map);
+    return Pair.of(required, ImmutableList.of(required.replace(newCollation)));
+  }
+
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> deriveTraits(
+      final RelTraitSet childTraits, final int childId) {
+    RelCollation collation = childTraits.getCollation();
+    if (collation == null || collation.getFieldCollations().size() == 0) {
+      return null;
+    }
+
+    return Pair.of(getTraitSet().replace(collation), ImmutableList.of(childTraits));
```

Comment:
This is not correct.
```
select distinct a, b*2, c from (select a, b, c from foo order by a, b, c limit 100) t;
```

Reply:
And it will be nice if we consider SqlMonotonicity here.

Reply:
Could you explain a bit on why this is not correct based on your query?

Reply:
The input is sorted by a, b, c.
The project is a, b*2, c
Based on this logic, the project is sorted by a, b*2, c. But in fact it is only sorted by a.

Reply:
Gotcha. I think your example query misses `order by a` at the end.

Full query
```
select distinct a, b*2, c 
 from (select a, b, c from foo order by a, b, c limit 100) t
order by a;
```
Let me think about it.

Reply:
No matter if has `order by a` or not, we need to see the plan:
```
SortedAgg
    -> Sort
           -> Project
                  -> Sort with limit
                             -> TableScan
```
In your current logic, the plan is:
```
SortedAgg
           -> Project
                  -> Sort with limit
                             -> TableScan
```

Reply:
And we'd better add another piece of logic to deal with SqlMonotonicity in traits derivation, similar with pass down.

Reply:
Have added the logic to deal with SqlMonotonicity for traits derivation, and add one positive test case `testSortProjectDeriveWhenCastLeadingToMonotonic` and one negative test case `testSortProjectDeriveWhenCastLeadingToNonMonotonic`

Reply:
And regarding the distinct query, I added it as the query in test case `testSortProjectDerive2`, where it generates a plan:

```
EnumerableSortedAggregate(group=[{0, 1, 2}])
  EnumerableSort(sort0=[$0], sort1=[$1], sort2=[$2], dir0=[ASC], dir1=[ASC], dir2=[ASC])
    EnumerableProject(ENAME=[$0], EXPR$1=[*($2, -2)], MGR=[$1])
      EnumerableLimit(fetch=[100])
        EnumerableProject(ENAME=[$1], MGR=[$3], SAL=[$5])
          EnumerableSort(sort0=[$1], sort1=[$3], sort2=[$5], dir0=[ASC], dir1=[ASC], dir2=[ASC])
            EnumerableTableScan(table=[[CATALOG, SALES, EMP]])
```

Reply:
Correct, this is expected. 

Can you add some comments to state the purpose of these tests. Sorry for not giving a good example in my code.

Reply:
That's a great point. I also realized new tests's names cannot explain their purpose already. I was thinking a better way.

Will add more details comments.